data-node/src/main/java/org/graylog/datanode/docs/ConfigurationDocsGenerator.java

```diff
@@ -0,0 +1,178 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog.datanode.docs;
+
+import com.github.joschi.jadconfig.Parameter;
+import com.github.joschi.jadconfig.ReflectionUtils;
+import jakarta.annotation.Nonnull;
+import org.apache.commons.lang3.ClassUtils;
+import org.graylog.datanode.docs.printers.ConfigFileDocsPrinter;
+import org.graylog.datanode.docs.printers.ConfigurationSection;
+import org.graylog.datanode.docs.printers.CsvDocsPrinter;
+import org.graylog.datanode.docs.printers.DocsPrinter;
+import org.graylog2.configuration.Documentation;
+import org.graylog2.configuration.DocumentationSection;
+
+import java.io.FileWriter;
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Locale;
+import java.util.Optional;
+import java.util.ServiceLoader;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+public class ConfigurationDocsGenerator {
+
+    /**
+     * This class is linked from the datanode pom.xml and generates conf.example and csv documentation.
+     */
+    public static void main(String[] args) throws IOException {
+        final ConfigurationDocsGenerator generator = new ConfigurationDocsGenerator();
+        generator.generateDocumentation(parseDocumentationFormat(args), ConfigurationBeansSPI::loadConfigurationBeans);
+    }
+
+    @Nonnull
+    private static DocumentationFormat parseDocumentationFormat(String[] args) {
+        if (args.length != 2) {
+            throw new IllegalArgumentException("This command needs two arguments - a format and file path. For example" + "csv ${project.build.directory}/configuration-documentation.csv");
+        }
+        final String format = args[0].toLowerCase(Locale.ROOT).trim();
+        final String file = args[1];
+        return new DocumentationFormat(format, file);
+    }
+
+    public void generateDocumentation(DocumentationFormat format, Supplier<List<Object>> configurationBeans) throws IOException {
+        final List<ConfigurationSection> sections = detectConfigurationSections(configurationBeans);
+        try (final DocsPrinter writer = createWriter(format)) {
+            writer.write(sections);
+        }
+    }
+
+    private DocsPrinter createWriter(DocumentationFormat format) throws IOException {
+        final FileWriter fileWriter = new FileWriter(format.outputFile(), StandardCharsets.UTF_8);
+        return switch (format.format()) {
```

Comment:
this is factory logic and should maybe be moved to a real factory class

Reply:
I'd wait if we will want to extend the logic and then maybe refactor. Added a TODO there. If we'll implement a third writer, then it will be for sure better to refactor.